Firestore/core/src/firebase/firestore/local/leveldb_transaction.cc

```diff
@@ -0,0 +1,196 @@
+/*
+ * Copyright 2018 Google
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "Firestore/core/src/firebase/firestore/local/leveldb_transaction.h"
+
+#include <leveldb/write_batch.h>
+#if __OBJC__
+#import <Protobuf/GPBProtocolBuffers.h>
+#endif
+
+#include "Firestore/core/src/firebase/firestore/util/firebase_assert.h"
+
+using leveldb::DB;
+using leveldb::ReadOptions;
+using leveldb::Slice;
+using leveldb::Status;
+using leveldb::WriteBatch;
+using leveldb::WriteOptions;
+
+namespace firebase {
+namespace firestore {
+namespace local {
+
+LevelDBTransaction::LevelDBTransaction(std::shared_ptr<DB> db,
+                                       const ReadOptions& readOptions,
+                                       const WriteOptions& writeOptions)
+    : db_(db), readOptions_(readOptions), writeOptions_(writeOptions) {
+}
+
+void LevelDBTransaction::Put(const std::string& key, const std::string& value) {
+  mutations_[key] = value;
+  deletions_.erase(key);
+  version_++;
+}
+
+LevelDBTransaction::Iterator::Iterator(LevelDBTransaction* txn)
+    : ldb_iter_(txn->db_->NewIterator(txn->readOptions_)),
+      txn_(txn),
+      last_version_(txn->version_),
+      is_valid_(false),  // Iterator doesn't really point to anything yet, so is
+                         // invalid
+      mutations_iter_(
+          std::make_unique<Mutations::iterator>(txn->mutations_.begin())) {
+}
+
+void LevelDBTransaction::Iterator::UpdateCurrent() {
+  bool mutation_is_valid = *mutations_iter_ != txn_->mutations_.end();
+  is_valid_ = mutation_is_valid || ldb_iter_->Valid();
+
+  if (is_valid_) {
+    if (!mutation_is_valid) {
+      is_mutation_ = false;
+    } else if (!ldb_iter_->Valid()) {
+      is_mutation_ = true;
+    } else {
+      // both are valid
+      const std::string mutation_key = (*mutations_iter_)->first;
+      const std::string ldb_key = ldb_iter_->key().ToString();
+      is_mutation_ = mutation_key <= ldb_key;
+    }
+    if (is_mutation_) {
+      current_.first = (*mutations_iter_)->first;
+      current_.second = (*mutations_iter_)->second;
+    } else {
+      current_.first = ldb_iter_->key().ToString();
+      current_.second = ldb_iter_->value().ToString();
+    }
+  }
+}
+
+void LevelDBTransaction::Iterator::Seek(const std::string& key) {
+  ldb_iter_->Seek(key);
+  for (; ldb_iter_->Valid() &&
+         txn_->deletions_.find(ldb_iter_->key().ToString()) !=
+             txn_->deletions_.end();
```

Comment:
You're repeating this incantation several times in this file this would be cleaner with a helper on transaction like:

```c++
bool IsDeleted(leveldb::Slice key) {
  return deletions_.find(key.ToString()) != deletions_.end();
}
```

Reply:
Done.